modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/IndexRebuildIntegrationTest.java

```diff
@@ -0,0 +1,316 @@
+import org.apache.ignite.internal.managers.indexing.IndexesRebuildTask;
+import org.apache.ignite.internal.processors.cache.GridCacheContext;
+import org.apache.ignite.internal.processors.query.calcite.QueryChecker;
+import org.apache.ignite.internal.processors.query.schema.IndexRebuildCancelToken;
+import org.apache.ignite.internal.processors.query.schema.SchemaIndexCacheVisitorClosure;
+import org.apache.ignite.internal.util.future.GridFutureAdapter;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.hamcrest.CoreMatchers;
+import org.junit.Test;
+
+/**
+ * Test index rebuild.
+ */
+public class IndexRebuildIntegrationTest extends AbstractBasicIntegrationTest {
+    /** Index rebuild init latch. */
+    private static CountDownLatch initLatch;
+
+    /** Index rebuild start latch. */
+    private static CountDownLatch startLatch;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IndexProcessor.idxRebuildCls = BlockingIndexesRebuildTask.class;
+
+        return super.getConfiguration(igniteInstanceName).setDataStorageConfiguration(
+            new DataStorageConfiguration().setDefaultDataRegionConfiguration(
+                new DataRegionConfiguration().setPersistenceEnabled(true)));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected int nodeCount() {
+        return 2;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        initLatch = null;
+        startLatch = null;
+
+        super.beforeTest();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() {
+        // Override super method to skip caches destroy.
+        cleanQueryPlanCache();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        cleanPersistenceDir();
+
+        super.beforeTestsStarted();
+
+        client.cluster().state(ClusterState.ACTIVE);
+
+        executeSql("CREATE TABLE tbl (id INT PRIMARY KEY, val VARCHAR, val2 VARCHAR) WITH CACHE_NAME=\"test\"");
+        executeSql("CREATE INDEX idx_id_val ON tbl (id DESC, val)");
+        executeSql("CREATE INDEX idx_id_val2 ON tbl (id, val2 DESC)");
+        executeSql("CREATE INDEX idx_val ON tbl (val DESC)");
+
+        for (int i = 0; i < 100; i++)
+            executeSql("INSERT INTO tbl VALUES (?, ?, ?)", i, "val" + i, "val" + i);
+
+        executeSql("CREATE TABLE tbl2 (id INT PRIMARY KEY, val VARCHAR)");
+
+        for (int i = 0; i < 100; i++)
+            executeSql("INSERT INTO tbl2 VALUES (?, ?)", i, "val" + i);
+    }
+
+    /** */
+    @Test
+    public void testRebuildOnInitiatorNode() throws Exception {
+        String sql = "SELECT * FROM tbl WHERE id = 0 AND val='val0'";
+
+        QueryChecker validChecker = assertQuery(grid(0), sql)
+            .matches(QueryChecker.containsIndexScan("PUBLIC", "TBL", "IDX_ID_VAL"))
```

Comment:
We need to stop checking query plans in integration tests. There is already a test IndexRebuildPlannerTest which verifies a planner behaviour during index rebuilding. For integration tests the only thing that matters is the result set.

Reply:
I disagree here. If the plan will be changed for some of these queries and there will be no index scan rewrite anymore, nobody will notice it, functionality became uncovered by tests and easily can be broken.

Reply:
This functionality is covered by the planner test. Besides, even with such a validation you still can't be sure you are verifying the right scenario, because for explain call the query will be optimized the second time, hence the plan could be different.

Reply:
Then we need to reuse the same queries for integration and planning tests. And we should ensure that queries in planning tests can't be changed without corresponding changes to integration tests. Perhaps, introduce some constants with queries. But I think it's not much convenient, so it's better to just recheck the required rel op in integration tests.

Without concurrent workload, the plan should be the same for explain and the query. It's much more likely that plan in planner test for the same query will be different than plan for integration test (and there are a lot of such cases currently).

Reply:
My point is you don't need to duplicate tests from planner test in integration test suite. And you don't need to duplicate test from LogicalRelImplementorTest too. Both planner part and rewriter part are covered.

Imo, for the integration part, we need to check if the cluster returns a valid result in the event that an index rebuild is triggered.

Reply:
It's not a duplication, LogicalRelImplementorTest doesn't check execution node internals (and can't check it without reflection or without exposing these internals).

The integration part does not check the plan, it checks result, but ensures that this result was received using the correct code path.